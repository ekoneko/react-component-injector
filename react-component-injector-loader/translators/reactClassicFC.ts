/**
 * React translate jsx with React.createElement before 17.
 * It need import react manually.
 */

import * as t from "@babel/types";
import { addInjectorLib, checkInjectorLibExists, wrapVariable } from "../utils";

export function getReactTag(programNode: t.Program) {
  const reactDeclarationNode = programNode.body.find(
    (node) =>
      node.type === "ImportDeclaration" &&
      node.source.value === "react"
  ) as t.ImportDeclaration | undefined;
  const reactDefaultSpecifier = reactDeclarationNode?.specifiers.find(specifier => specifier.type === 'ImportDefaultSpecifier') as t.ImportDefaultSpecifier | null
  return reactDefaultSpecifier?.local.name
}

export function wrapReactClassicFC(programNode: t.Program, { reactTag, resourceRelativePath, encrypt }: { reactTag: string; resourceRelativePath: string; encrypt: (key: string) => string }) {
  let injectorLibExists = checkInjectorLibExists(programNode);
  programNode.body.forEach((node, index) => {
    if (node.type !== "ExportNamedDeclaration") {
      return;
    }
    const variableDeclarationNode = node.declaration;
    if (
      variableDeclarationNode?.type !== "VariableDeclaration" ||
      variableDeclarationNode.declarations[0]?.type !==
      "VariableDeclarator" ||
      variableDeclarationNode.declarations[0]?.init?.type !==
      "FunctionExpression"
    ) {
      return;
    }
    const returnStatementNode =
      variableDeclarationNode.declarations[0]?.init.body.body.find(
        (item) => item.type === "ReturnStatement"
      ) as t.ReturnStatement;
    if (returnStatementNode?.argument?.type !== "CallExpression") {
      return;
    }
    if (returnStatementNode?.argument?.type !== "CallExpression") {
      return;
    }
    const returnCallee = returnStatementNode.argument.callee;
    if (
      returnCallee.type !== 'MemberExpression' ||
      returnCallee.object.type !== 'Identifier' ||
      returnCallee.object.name !== reactTag ||
      returnCallee.property.type !== 'Identifier' ||
      returnCallee.property.name !== 'createElement' ||
      variableDeclarationNode.declarations[0].id.type !== "Identifier"
    ) {
      return;
    }
    const componentName = variableDeclarationNode.declarations[0].id.name;
    if (!injectorLibExists) {
      addInjectorLib(programNode);
      index += 1;
      injectorLibExists = true;
    }
    programNode.body.splice(index, 1, variableDeclarationNode, ...wrapVariable(componentName, encrypt(`${resourceRelativePath}#${componentName}`)));
  });
}