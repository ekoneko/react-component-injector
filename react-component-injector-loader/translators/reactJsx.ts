/**
 * React-jsx is a new way to parse jsx code which not need import React.
 * On this scene, jsx file will auto import "react/jsx-runtime" 
 * and transform jsx to "_jsx("div", { children: "Text" })"
 */
import * as t from "@babel/types";
import { addInjectorLib, checkInjectorLibExists, wrapVariable } from '../utils'

export function getJsxRuntimeTag(programNode: t.Program) {
  const jsxRuntimeDeclarationNode = programNode.body.find(
    (node) =>
      node.type === "ImportDeclaration" &&
      node.source.value === "react/jsx-runtime"
  ) as t.ImportDeclaration | undefined;
  return (
    // react/jsx-runtime is added automated, the type never changed.
    (
      jsxRuntimeDeclarationNode?.specifiers as [t.ImportSpecifier] | undefined
    )?.[0].local.name
  );
}

export function wrapReactJsxComponent(programNode: t.Program, { jsxTag, resourceRelativePath, encrypt }: { jsxTag: string; resourceRelativePath: string; encrypt: (key: string) => string }) {
  let injectorLibExists = checkInjectorLibExists(programNode);
  programNode.body.forEach((node, index) => {
    if (node.type !== "ExportNamedDeclaration") {
      return;
    }
    const variableDeclarationNode = node.declaration;
    if (
      variableDeclarationNode?.type !== "VariableDeclaration" ||
      variableDeclarationNode.declarations[0]?.type !==
      "VariableDeclarator" ||
      variableDeclarationNode.declarations[0]?.init?.type !==
      "FunctionExpression"
    ) {
      return;
    }
    const returnStatementNode =
      variableDeclarationNode.declarations[0]?.init.body.body.find(
        (item) => item.type === "ReturnStatement"
      ) as t.ReturnStatement;
    if (returnStatementNode?.argument?.type !== "CallExpression") {
      return;
    }
    const returnCallee = returnStatementNode.argument.callee;
    if (
      returnCallee.type !== "Identifier" ||
      returnCallee.name !== jsxTag ||
      variableDeclarationNode.declarations[0].id.type !== "Identifier"
    ) {
      return;
    }
    const componentName = variableDeclarationNode.declarations[0].id.name;
    if (!injectorLibExists) {
      addInjectorLib(programNode);
      index += 1;
      injectorLibExists = true;
    }
    programNode.body.splice(index, 1, variableDeclarationNode, ...wrapVariable(componentName, encrypt(`${resourceRelativePath}#${componentName}`)));
  });
}