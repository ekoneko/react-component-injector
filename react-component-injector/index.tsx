import React from "react";
import md5 from "md5";

interface InjectedOptions<P = any> {
  originComponentType: React.ComponentType<P>;
}

type InjectedFn<P = any> = (
  options: InjectedOptions<P>
) => React.ComponentType<P>;

class ReactComponentInjector {
  injectedComponentsMap: Record<string, InjectedFn> = {};

  proxy<P extends object>(
    key: string,
    OriginComponentType: React.ComponentType<P>
  ) {
    const Wrapped: React.FC<P> = (props) => {
      const injectedFn = this.injectedComponentsMap[key];
      const ComponentType =
        injectedFn?.({ originComponentType: OriginComponentType }) ??
        OriginComponentType;
      return <ComponentType {...props} />;
    };
    return Wrapped;
  }

  inject<P extends React.ComponentType<any>>(
    path: string,
    injectedFn: InjectedFn<P>
  ) {
    const key = this.encryptPath(path);
    this.injectedComponentsMap[key] = injectedFn;
  }

  private encryptPath(key: string) {
    // NOTE: use same way as loader
    return md5(key);
  }
}
export const reactComponentInjector = new ReactComponentInjector();
