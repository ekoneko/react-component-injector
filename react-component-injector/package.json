{
  "name": "react-component-injector",
  "version": "1.0.0",
  "main": "./lib/index.js"
}