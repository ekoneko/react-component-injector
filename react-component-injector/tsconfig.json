{
  "extends": "../tsconfig.base.json",
  "compilerOptions": {
    "jsx": "react-jsx",
    "lib": ["esnext"],
    "outDir": "lib",
    "declaration": true
  },
  "files": ["index.tsx"]
}
