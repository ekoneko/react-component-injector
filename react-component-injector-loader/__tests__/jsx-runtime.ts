import { parse as babelParse } from "@babel/parser";
import reactInjectorLoader from '../index'
import { getJsxRuntimeTag } from "../translators/reactJsx";

const code = `import { jsx as _jsx } from "react/jsx-runtime";
import { reactComponentInjector } from "react-component-injector";
const a = 'test';
export var A = function () { var a = 'xxx'; return _jsx("span", { children: "Hello" }); };

var B = function () { return _jsx("span", { children: "World" }); };
var WrappedB = reactComponentInjector.define("src/ComponentB#B", B);
export { WrappedB as B };`

describe('jsx-runtime', () => {
  test('detect', () => {
    const ast = babelParse(code, { sourceType: "module" });
    expect(getJsxRuntimeTag(ast.program)).toEqual('_jsx')
  })
  test('wrap', () => {
    expect(reactInjectorLoader.bind({
      resourcePath: '/path/to/component.tsx',
      getOptions: () => ({
        encrypt: (key: string) => key
      })
    })(code)).toMatchSnapshot();
  })
})