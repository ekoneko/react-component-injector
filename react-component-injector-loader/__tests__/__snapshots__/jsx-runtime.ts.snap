// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsx-runtime 1`] = `
"import { jsx as _jsx } from \\"react/jsx-runtime\\";
import { reactComponentInjector } from \\"react-component-injector\\";
const a = 'test';

var A = function () {
  var a = 'xxx';
  return _jsx(\\"span\\", {
    children: \\"Hello\\"
  });
};

var $$WrappedA = reactComponentInjector.proxy(\\"/path/to/component.tsx#A\\", A);
export { $$WrappedA as A };

var B = function () {
  return _jsx(\\"span\\", {
    children: \\"World\\"
  });
};

var WrappedB = reactComponentInjector.define(\\"src/ComponentB#B\\", B);
export { WrappedB as B };"
`;

exports[`jsx-runtime wrap 1`] = `undefined`;
